### PR TITLE
Fix kexec framebuffer graphics for 5.10 kernel

### PR DESCRIPTION
**Pull request by JonathonHall-Purism**
Fix setting up a framebuffer for the OS kernel to use following kexec.

If Heads is configured with kernel 4.20 or later, the OS kernel is not able to use framebuffer video output.  The OS kernel must have the i915 module (or other appropriate driver) and there won't be any video output until it takes over the device.  In particular, this prevents booting memtest86+ and the Debian netinstall image (as well as others).

This now works on i915 using kernel 5.10 in Heads.  I've tested Librem 15v4 and Librem Mini v2.

It does not work on Aspeed graphics (ast), even when adding the astdrmfb identifier to kexec.  It looks like ast uses a shadow buffer internally in the kernel (it sets `mode_config.prefer_shadow = 1` in ast_mode.c), which means that the generic DRM FB helper doesn't know where the real framebuffer is (AFAICT from a cursory read-through).  A hack would be to somehow get the real framebuffer address up to fix.smem_start, but this is pretty wrong from the kernel's perspective (nothing should write there, it's just what we need since we are going to kexec and the driver will be gone), would just be a hack for kexec.  I took an initial look at this but it's going to need a deeper dive.

That's probably the reason for the ID checks in kexec, so I don't think we can remove that currently.

---

TODO:

- [x] Test more boards, particularly Aspeed graphics (server)
- [x] Bump kexec to 2.0.26
- [x] See if we can remove the driver name checks in kexec
- [x] Improve kexec tracing, in particular trace unknown driver names (if we have to keep that check) and warn if framebuffer address is not provided by kernel
- [x] Enable fixes for other boards

### CODE REVIEW

**Comment by tlaurion** [comment]
I suggest that we version bump kexec to latest.

I added the following patch on my debugging branch so that we know what is happening when calling `kexec -d -l` with debugging.

```
diff --git a/Makefile.in b/Makefile.in
index fb01134..bf1973e 100644
--- a/Makefile.in
+++ b/Makefile.in
@@ -157,12 +157,12 @@ include $(srcdir)/kexec/Makefile
 
 # vmcore-dmesg (read dmesg from a vmcore)
 #
-include $(srcdir)/vmcore-dmesg/Makefile
+#include $(srcdir)/vmcore-dmesg/Makefile
 
 #
 # kexec_test (test program)
 #
-include $(srcdir)/kexec_test/Makefile
+#include $(srcdir)/kexec_test/Makefile
 
 SPEC=$(PACKAGE_NAME).spec
 GENERATED_SRCS:= $(SPEC)
diff --git a/kexec/arch/i386/x86-linux-setup.c b/kexec/arch/i386/x86-linux-setup.c
index 057ee14..43e017a 100644
--- a/kexec/arch/i386/x86-linux-setup.c
+++ b/kexec/arch/i386/x86-linux-setup.c
@@ -137,7 +137,8 @@ static int setup_linux_vesafb(struct x86_linux_param_header *real_mode)
 		goto out;
 	if (-1 == ioctl(fd, FBIOGET_VSCREENINFO, &var))
 		goto out;
-	if (0 == strcmp(fix.id, "VESA VGA")) {
+	if (0 == strcmp(fix.id, "VESA VGA")
+            || 0 == strcmp(fix.id, "inteldrmfb") || 0 == strcmp(fix.id, "i915drmfb")) {
 		/* VIDEO_TYPE_VLFB */
 		real_mode->orig_video_isVGA = 0x23;
 	} else if (0 == strcmp(fix.id, "EFI VGA")) {
diff --git a/kexec/kexec.c b/kexec/kexec.c
index bc6ab3d..b82725b 100644
--- a/kexec/kexec.c
+++ b/kexec/kexec.c
@@ -805,6 +805,27 @@ static int my_load(const char *type, int fileind, int argc, char **argv,
 	if (sort_segments(&info) < 0) {
 		return -1;
 	}
+
+#if 1
+	// force segment 0 to have memsz == bufsz
+	// so that it won't overwrite EBDA
+	if (info.segment[0].mem == 0)
+	{
+		if (kexec_debug)
+			printf("hack ebda into segment 0!\n");
+
+		uint8_t * ebda = calloc(1, info.segment[0].memsz);
+		memcpy(ebda, info.segment[0].buf, info.segment[0].bufsz);
+		info.segment[0].bufsz = info.segment[0].memsz;
+		info.segment[0].buf = ebda;
+
+		// install some default EBDA values that are off scale,
+		// which will force Xen to use the multiboot info
+		*(uint16_t*)(ebda + 0x40e) = 0xFFFF; // segment
+		*(uint16_t*)(ebda + 0x413) = 0xFFFF; // size
+	}
+#endif
+
 	/* if purgatory is loaded update it */
 	update_purgatory(&info);
 	if (entry)
diff --git a/purgatory/Makefile b/purgatory/Makefile
index 2dd6c47..2de8f07 100644
--- a/purgatory/Makefile
+++ b/purgatory/Makefile
@@ -44,7 +44,6 @@ purgatory/sha256.o: $(srcdir)/util_lib/sha256.c
 	mkdir -p $(@D)
 	$(COMPILE.c) -o $@ $^
 
-$(PURGATORY): CC=$(TARGET_CC)
 $(PURGATORY): CFLAGS=$(PURGATORY_EXTRA_CFLAGS) \
 		      $($(ARCH)_PURGATORY_EXTRA_CFLAGS) \
 		      -Os -fno-builtin -ffreestanding \
diff --git a/util/Makefile b/util/Makefile
index 948ee63..833a897 100644
--- a/util/Makefile
+++ b/util/Makefile
@@ -2,7 +2,7 @@ BIN_TO_HEX:= bin/bin-to-hex
 
 $(BIN_TO_HEX): $(srcdir)/util/bin-to-hex.c
 	@$(MKDIR) -p $(@D)
-	$(LINK.o) $(CFLAGS) -o $@ $^
+	$(BUILD_CC) $(BUILD_CFLAGS) -o $@ $^
 
 $(BIN_TO_HEX): CC=$(BUILD_CC)
 $(BIN_TO_HEX): CFLAGS=$(BUILD_CFLAGS)
--- a/kexec/arch/i386/x86-linux-setup.c.orig	2023-04-11 15:44:59.905000000 -0400
+++ b/kexec/arch/i386/x86-linux-setup.c	2023-04-11 15:45:03.830000000 -0400
@@ -146,9 +146,11 @@
             || 0 == strcmp(fix.id, "inteldrmfb") || 0 == strcmp(fix.id, "i915drmfb")) {
 		/* VIDEO_TYPE_VLFB */
 		real_mode->orig_video_isVGA = 0x23;
+		dbgprintf("%s: VIDEO_TYPE_VLFB setuped.\n", __func__);
 	} else if (0 == strcmp(fix.id, "EFI VGA")) {
 		/* VIDEO_TYPE_EFI */
 		real_mode->orig_video_isVGA = 0x70;
+		dbgprintf("%s: VIDEO_TYPE_EFI setuped.\n", __func__);
 	} else if (arch_options.reuse_video_type) {
 		int err;
 		off_t offset = offsetof(typeof(*real_mode), orig_video_isVGA);
@@ -159,6 +161,7 @@
 			goto out;
 	} else {
 		real_mode->orig_video_isVGA = 0;
+		dbgprintf("%s: VIDEO_TYPE not VGA.\n", __func__);
 		close(fd);
 		return 0;
 	}

Might be a good idea to output fix.id content when no match happens into kexec. Otherwise we are kinda in the dark right now
```

**Comment by JonathonHall-Purism** [comment]
The big picture of how boot works is somewhat complex.  In particular, it depends on whether coreboot provides any graphics init and a framebuffer that the kernel can observe (such as VBE).  My boards don't provide anything from coreboot, and we would prefer not to have to include libgfxinit since we are already shipping a driver in the Heads kernel.

qemu works differently and I have not investigated that yet, I'm focusing on hardware first.

The boot is supposed to proceed this way:
* coreboot invokes the Heads kernel.  No graphics are initialized yet.
* Heads' kernel uses the i915 driver to select a mode and create a framebuffer.
* When handing off to the OS kernel, kexec(8) fills in the OS kernel boot parameters (since it will not invoke the OS kernel's real mode stub).
* kexec(8) uses the Heads kernel's framebuffer to fill in the video mode and framebuffer for the OS kernel.  (This framebuffer was set up by i915, but the OS kernel doesn't know that and will hand it to the vesafb driver, which uses a flat framebuffer.)
* The OS kernel uses that framebuffer for early console.
* If the OS has i915, it'll eventually take over.  Otherwise it uses the framebuffer forever.

The upshot of all this is that we can boot an OS needing a framebuffer without having to ship anything like a VBIOS, GOP driver, etc.  We just use the Heads Linux kernel to set up a framebuffer.

There's been a lot of discussion on this in a lot of places, and just to capture a few more details:
* kexec --reuse-video-type is not useful to us.  This would copy whatever was in the Heads kernel's boot parameters to the OS.  This is probably "nothing" because coreboot is not doing any graphics init, and even if it was something, i915 has changed modes by this point.
* There's some confusion about what happens in real mode (16 bit), protected mode (32 bit), long mode (64 bit) in the OS kernel.  VBE is probed in real mode, which doesn't happen here (we don't invoke the real mode stub, and there is no VBE to detect).  The OS kernel says it is using "vesafb", but that's just the driver it uses to handle a flat framebuffer, it doesn't actually know where the framebuffer came from.

**Comment by JonathonHall-Purism** [comment]
> I suggest that we version bump kexec to latest.

Agree, I'm working locally with 2.0.26, right now just want to get it functioning correctly and then we can expand with things like this.

> Might be a good idea to output fix.id content when no match happens into kexec. Otherwise we are kinda in the dark right now

Agree here too, I had to do the same to get the new-new ID.  But maybe we can improve it not to check ID this way, since it's a bit of a cat-and-mouse game keeping up with the ID changes.  Anyway, like above, I want to get it to fully work on 5.10, try other boards, and see what we can improve at that point.

**Comment by daringer** [comment]
As to be seen by the reference, this solution works for me on a novacustom nv4xpz (12th gen intel) together with some more twirks to get into heads initially (simplefb). Here's my current state: https://github.com/osresearch/heads/compare/master...daringer:heads:nitropad-x 

Notable changes/differences: GOP init for gfx + simplefb + 6.1 kernel

awesome work @JonathonHall-Purism thanks!

**Comment by tlaurion** [comment]
[b7adbb6](https://github.com/tlaurion/heads/commit/b7adbb6cdccd13847a968ea07ff7862d59b9dc1c) x230-maximized rom was able to boot directly into Tinycore 64 bit (http://tinycorelinux.net/14.x/x86_64/release/TinyCorePure64-14.0.iso) dd'ed onto usb thumb drive. (@saber might be interested. Other isolinux fix are needed to boot detached signed iso, which is #1374 )

**Comment by paulmenzel** [comment]
> If Heads is configured with kernel 4.20 or later, the OS kernel is not able to use framebuffer video output.

Was some configuration changed? If not, it’d be great, if you bisected this regression.

**Comment by paulmenzel** [comment]
Was Linux’ framebuffer configuration changed? Does building with simplefb fix it as it’s done in @daringer’s changes?

>    CONFIG_SYSFB_SIMPLEFB=y
>    CONFIG_FB=y

**Comment by JonathonHall-Purism** [comment]
@paulmenzel I did bisect this (twice :grin: )

The first problem was this commit: https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/commit/?id=6be8f3bd2c78915a9f3a058a346ae93068d35c01

Linux now hides the framebuffer address from userspace.  That makes sense, but kexec uses it to populate the boot parameters for the new kernel.  There were a Kconfig and module parameter added later to keep the old behavior for compatibility, which I enabled here.

The second problem (specific to i915 apparently) was this: https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/commit/?id=691f7ba58d5220bbb06392054a6e50abdd393516

More devices started using a compressed framebuffer in more situations.  We can only provide a flat framebuffer to the new OS, so I set the module parameter to disable this.

I'm not really sure what the long term "perfect solution" is here, it seems like having kexec do this is making a lot of assumptions about how the kernel has set this framebuffer up, and it still doesn't work on ast (and probably others), looks like they are using a shadow framebuffer in the kernel and we need the real framebuffer address somehow.  This works for i915 for now, but maybe a more ideal solution would be to have the host kernel provide its framebuffer address to the new kernel instead of doing it in userspace?

>CONFIG_SYSFB_SIMPLEFB=y

Don't think so, because coreboot isn't initializing graphics.  The first graphics initialization is done by the Heads kernel's i915 driver (or other driver), then we use that as the "firmware" framebuffer for the new kernel.

>CONFIG_FB=y

This one is already enabled.

**Comment by daringer** [comment]
> > CONFIG_SYSFB_SIMPLEFB=y
> 
> Don't think so, because coreboot isn't initializing graphics. The first graphics initialization is done by the Heads kernel's i915 driver (or other driver), then we use that as the "firmware" framebuffer for the new kernel.

For my hardware setup this was the only thing that helped for HEADS to display anything at all. But I did not try to leave it out and use your kernel commandline instead:
```
CONFIG_LINUX_COMMAND_LINE="iommu=pt quiet loglevel=2 video=eDP-1:1920x1080 drm_kms_helper.drm_leak_fbdev_smem=1 i915.enable_fbc=0"
```
@JonathonHall-Purism do you think using this command line instead of `CONFIG_SYSFB_SIMPLEFB` should work? and if so,  does it make any difference ?

**Comment by daringer** [comment]
> > CONFIG_SYSFB_SIMPLEFB=y
> 
> Don't think so, because coreboot isn't initializing graphics. The first graphics initialization is done by the Heads kernel's i915 driver (or other driver), then we use that as the "firmware" framebuffer for the new kernel.

For my hardware setup this was the only thing that helped for HEADS to display anything at all. But I did not try to leave it out and use your kernel commandline instead:
```
CONFIG_LINUX_COMMAND_LINE="iommu=pt quiet loglevel=2 video=eDP-1:1920x1080 drm_kms_helper.drm_leak_fbdev_smem=1 i915.enable_fbc=0"
```
@JonathonHall-Purism do you think using this command line instead of `CONFIG_SYSFB_SIMPLEFB` should work? and if so,  does it make any difference ?

**Comment by JonathonHall-Purism** [comment]
@tlaurion I enabled the fixes for the other 5.10/i915 boards - other Librems except L1UM, and t440p.  I tested mini v2 (I'll test the others in my next PB release).  Could you test t440p?

Aspeed graphics still do not work per the edited OP, I don't think I will be able to dive into this right away, I think this is worth finishing for i915 boards and merging.  That won't be as simple (hah) of a fix since the real framebuffer address isn't plumbed near anywhere kexec could get it right now, as far as I can tell.

I don't think we can drop the ID checks right now as a result.  Maybe we could check if the smem_start is nonzero instead, not sure if ast is populating the shadow framebuffer here or nothing.

I will still bump kexec to 2.0.26, this should be easy, and grab your improvements to tracing, particularly the driver name.

**Comment by JonathonHall-Purism** [comment]
> @JonathonHall-Purism do you think using this command line instead of `CONFIG_SYSFB_SIMPLEFB` should work? and if so, does it make any difference ?

It depends a bit on how graphics are initialized on your board.  If it's like mine (nothing from coreboot, initialized by i915 in Heads kernel), then you need the kernel command line _and_ config changes in this PR.  CONFIG_SYSFB_SIMPLEFB will not help because there is no framebuffer provided by coreboot.

If, on the other hand, you _do_ have graphics initialized by coreboot, then you could try using simplefb and having kexec blindly reuse the video mode with --reuse-video-type.  You'd also need to remove the graphics driver from the Heads kernel (i915 or other driver) so it doesn't change modes, which would invalidate the firmware-provided mode I believe.

I don't really want to go that route on my boards though - I don't think libgfxinit supports them all so I would have to use VBIOS or a GOP driver (more blobs), and I'm not sure that would work on the servers (don't know if Aspeed's blobs would work without the full BIOS/UEFI environment).

**Comment by JonathonHall-Purism** [comment]
@tlaurion Updated to kexec 2.0.26 and added tracing.  I think I captured everything you had, and added a few things too.

If there is anything that I missed that you think should be in here, let me know.

Aspeed and qemu still aren't working but I don't really have bandwidth for another deep dive on this at the moment.  There's no regression here, so I think the i915 work can be merged, no reason to hold it.

**Comment by tlaurion** [comment]
@ThePlexus @srgrint @akunterkontrolle : I would need one of you confirming that:
- [ ] t440p works with current fix (it should!)

**Comment by srgrint** [comment]
I can confirm that I have flashed 353e836 on my t440p.  The debian netinst from USB (I have tried bookworm test image) now boots fine.  My installed debian distribution on the hard drive still works as well!

**Comment by JonathonHall-Purism** [comment]
This comment sounds like a blocker (thank you @lankredotp41 for mentioning it!): https://github.com/osresearch/heads/pull/1351#issuecomment-1517475981

Per the comment, newer kexec now prefers multiboot2 by default, defeating the segment 0 / EBDA patch.

We should check this out - @tlaurion the patch mentions Xen but I'm not sure if that whole chunk relates to Xen or just those few lines.  Do you know offhand what this patch was for?

**Comment by tlaurion** [comment]
@JonathonHall-Purism :
> My boards don't provide anything from coreboot, and we would prefer not to have to include libgfxinit since we are already shipping a driver in the Heads kernel.

We discussed it off github : coreboot savedefconfig format hides the facts, but mostly all librem boards currently enable libgfxinit and can now be more easily seen with helpers I added under https://github.com/osresearch/heads/pull/1381/commits/29c4b5ce52a7110d6545224665140a62ae8071ba.

```

user@heads-tests:~/heads$ find boards/ -type d | awk -F "/" {'print $2'}| while read board; do make BOARD=$board coreboot.save_in_oldconfig_format_in_place; done; grep -Rn "CONFIG_MAINBOARD_USE_LIBGFXINIT=y" config/coreboot-*
config/coreboot-librem_13v2.config:449:CONFIG_MAINBOARD_USE_LIBGFXINIT=y
config/coreboot-librem_13v4.config:449:CONFIG_MAINBOARD_USE_LIBGFXINIT=y
config/coreboot-librem_14.config:449:CONFIG_MAINBOARD_USE_LIBGFXINIT=y
config/coreboot-librem_15v3.config:449:CONFIG_MAINBOARD_USE_LIBGFXINIT=y
config/coreboot-librem_15v4.config:449:CONFIG_MAINBOARD_USE_LIBGFXINIT=y
config/coreboot-librem_mini.config:445:CONFIG_MAINBOARD_USE_LIBGFXINIT=y
config/coreboot-librem_mini_v2.config:445:CONFIG_MAINBOARD_USE_LIBGFXINIT=y
config/coreboot-t420.config:449:CONFIG_MAINBOARD_USE_LIBGFXINIT=y
config/coreboot-t420-maximized.config:450:CONFIG_MAINBOARD_USE_LIBGFXINIT=y
config/coreboot-t430-legacy.config:442:CONFIG_MAINBOARD_USE_LIBGFXINIT=y
config/coreboot-t430-maximized.config:449:CONFIG_MAINBOARD_USE_LIBGFXINIT=y
config/coreboot-t440p.config:388:CONFIG_MAINBOARD_USE_LIBGFXINIT=y
config/coreboot-t520-maximized.config:446:CONFIG_MAINBOARD_USE_LIBGFXINIT=y
config/coreboot-t530-maximized.config:452:CONFIG_MAINBOARD_USE_LIBGFXINIT=y
config/coreboot-w530-maximized.config:452:CONFIG_MAINBOARD_USE_LIBGFXINIT=y
config/coreboot-x230-maximized-fhd_edp.config:448:CONFIG_MAINBOARD_USE_LIBGFXINIT=y
```

**Comment by daringer** [comment]
Did someone test this with QubesOS, the same config which properly boots a recent Ubuntu is not working for me for QubesOS. It looks like the same issue we had with the linux kernel before. Using this https://github.com/daringer/heads/tree/nitropad-x setup... 

@marmarek are you maybe aware of issues with kexec with recent kernels + QubesOS/XEN ?

**Comment by daringer** [comment]
> Using this https://github.com/daringer/heads/tree/nitropad-x setup...

rebased on top of this PR, pulled in missing patches - I can confirm that using the updated kexec leads to no issues. So booting an Ubuntu LTS looks good, QubesOS still doesn't work but I suspect a too old kernel version here

**Comment by tlaurion** [comment]
@daringer as discussed off gihub:

https://www.qubes-os.org/news/2023/03/14/qubes-4-1-2/#whats-new-in-qubes-412

> kernel-latest available as a boot option when starting the installer

You have tried this option on installer?

**Comment by tlaurion** [comment]
> rebased on top of this PR, pulled in missing patches - I can confirm that using the updated kexec leads to no issues. So booting an Ubuntu LTS looks good, QubesOS still doesn't work but I suspect a too old kernel version here

Also discussed off github and above, kexec'ing into other kernel is not using the same codepath as multiboot (kexec 2.0.22) or multiboot2 (kexec 2.0.26). 

no-real-mode and vga=current are passed as additional arguments to Xen and kexec has a parch in tree that works for multiboot but not multiboot2.

I have tested booting q4.1 and q4.2 from this PR on x230. But graphics only appears after OS loads i915 driver.

Short version: i915drmfb additional I'd added into kexec patch affects non-multiboot kexec only: I have seen no improvement nor regression as compared to master when booting q4.1/q4.2(xen) whenever multiboot (master) or multiboot2 (this PR with kexec 2.0.26).

**Comment by JonathonHall-Purism** [comment]
I did some more testing on Mini v2 to clarify the behavior around coreboot libgfxinit.

Bottom line, yes we can CONFIG_NO_GFX_INIT, and I'll set it on these boards once I test a few more devices.

@tlaurion You're right that Librem boards actually did get libgfxinit by default, but the default is to initialize in text mode, so there was no framebuffer provided to the kernel like I had found earlier.  Disabling i915 booted in text mode.  CONFIG_X86_SYSFB (predecessor to CONFIG_SYSFB_SIMPLEFB) doesn't help either, which is not surprising since the output is in text mode.

I tried configuring coreboot to initialize in framebuffer mode instead (CONFIG_GENERIC_LINEAR_FRAMEBUFFER) but this seems to just fail on this board, the video output was not enabled at all.  There was no obvious tracing in coreboot indicating any error.  I didn't look into it further since this is not the direction we want to go anyway.

I'll test Qubes too.

**Comment by JonathonHall-Purism** [comment]
Tested 13v2, 14, and Mini v2, memtest86+ and PureOS work.  I can reproduce the issue with Qubes though; installed 4.1.2 on my L14 again and get no video on boot.  I'll need to investigate what is going on there.

**Comment by JonathonHall-Purism** [comment]
Well, I no longer have the issue with Qubes, not sure what was causing it.  I set it up on Mini v2 to get serial logs.  First boot after set up appeared to have no video output, but after about 5 minutes it actually did bring up video with i915.  After that every boot has been fine, and I tried installing again to see if it was a first-boot issue, that was fine too.  The install that was on my L14 is fine now too.  (Still only have graphics after i915 init, but that's the same as master.)

Unfortunately I didn't get a chance to collect any diagnostics since I wasn't expecting it to disappear on its own.  Given the circumstances, this may not be introduced by this PR.

@daringer Are you still having issues booting Qubes?

**Comment by daringer** [comment]
> @daringer Are you still having issues booting Qubes?

Yes, I still do. Are you using exactly this PR? I am afraid my problem might be a too old kernel version. But you say you installed QubesOS, already this is not working for me, once I kexec into the usb-device with the QubesOS image my display doesn't show anything.

I will clean up my working state and PR it here, this will then be easier to (and a better place to) discuss this.

**Comment by marmarek** [comment]
On Thu, Apr 27, 2023 at 02:19:31PM -0700, Markus Meissner wrote:
> Yes, I still do. Are you using exactly this PR? I am afraid my problem might be a too old kernel version. But you say you installed QubesOS, already this is not working for me, once I kexec into the usb-device with the QubesOS image my display doesn't show anything.

Are you doing kexec with -latest kernel as dom0, or the default one?

-- 
Best Regards,
Marek Marczykowski-Górecki
Invisible Things Lab

**Comment by daringer** [comment]
> Are you doing kexec with -latest kernel as dom0, or the default one?

Nope, I didn't - or to be more precise: I thought I did. Anyways it now works, the issue was self-inflicted as we removed the 6.1-kernel option from our QubesOS OEM images. Shot my own feet, both at once :nerd_face: 

@JonathonHall-Purism so I can confirm that there are no issues with QubesOS for this PR

**Comment by tlaurion** [comment]
> Tested 13v2, 14, and Mini v2, memtest86+ and PureOS work. I can reproduce the issue with Qubes though; installed 4.1.2 on my L14 again and get no video on boot. I'll need to investigate what is going on there.

@marmarek do you recall tests where qubesos would give early kernel console output prior if DRM driver kicking in?

**Comment by marmarek** [comment]
> @marmarek do you recall tests where qubesos would give early kernel console output prior if DRM driver kicking in?

Not really. That might have been the case before (Qubes 4.0?), but nor my memory nor test logs span that long to confirm/deny.

**Comment by JonathonHall-Purism** [comment]
Thanks @daringer and @marmarek :100: 

In that case @tlaurion I think this is ready to merge - fixes i915 framebuffer graphics for Linux distros and memtest.  Future work will be getting that to work on Aspeed as well, and possibly Qubes/Xen, but there's no regression, they're the same as before.

**Comment by tlaurion** [comment]
Haven't finished pushing 5.x config changes to xx20/xx30 yet but since they aren't impacted as of today, not a real problem

**Comment by akunterkontrolle** [comment]
Can confirm as well for the t440p that this pull request fixes the issues with no framebuffer graphical output. More detailed comment: https://github.com/osresearch/heads/issues/1323#issuecomment-1528718761
